src/Common/Identifiers.hs

```diff
@@ -49,106 +105,84 @@ instance ToIdent Identifier where
   deriving IsString via Identifier
   deriving Identifiable via Identifier
   deriving Semigroup via Identifier
   deriving Monoid via Identifier
   deriving Pretty via Identifier
 
-instance Show TConId where
-  show (TConId i) = "Ty" ++ show i
-
--- | ToIdentifier for type variable, e.g., @a@
+-- | ToIdentifier for type variable, e.g., @a@.
 newtype TVarId = TVarId Identifier
   deriving Eq
+  deriving Show
   deriving Ord
+  deriving Typeable
+  deriving Data
   deriving ToIdent via Identifier
   deriving IsString via Identifier
   deriving Identifiable via Identifier
   deriving Semigroup via Identifier
   deriving Monoid via Identifier
   deriving Pretty via Identifier
 
-instance Show TVarId where
-  show (TVarId i) = "tyvar" ++ show i
-
--- | de Bruijn index for type variables, e.g., @'0@
-newtype TVarIdx = TVarIdx Int
-  deriving Eq
-  deriving Ord
-
-instance Show TVarIdx where
-  show (TVarIdx i) = "'" ++ show i
-
-instance Pretty TVarIdx where
-  pretty = pretty . show
-
--- | ToIdentifier for data constructors, e.g., @None@
+-- | Identifier for data constructors, e.g., @None@.
 newtype DConId = DConId Identifier
   deriving Eq
+  deriving Show
   deriving Ord
-  deriving Show via Identifier
-  deriving ToIdent via Identifier
-  deriving IsString via Identifier
-  deriving Identifiable via Identifier
-  deriving Semigroup via Identifier
-  deriving Monoid via Identifier
-  deriving Pretty via Identifier
-
--- | ToIdentifier for low-level identifiers, e.g., @ssm_activate@
-newtype FfiId = FfiId Identifier
-  deriving Eq
-  deriving Ord
-  deriving Show via Identifier
+  deriving Typeable
+  deriving Data
   deriving ToIdent via Identifier
   deriving IsString via Identifier
   deriving Identifiable via Identifier
   deriving Semigroup via Identifier
   deriving Monoid via Identifier
   deriving Pretty via Identifier
 
--- | ToIdentifier for user-defined variable, e.g., @x@
+-- | Identifier for data variables, e.g., @x@.
 newtype VarId = VarId Identifier
   deriving Eq
+  deriving Show
   deriving Ord
-  deriving Show via Identifier
+  deriving Typeable
+  deriving Data
   deriving ToIdent via Identifier
   deriving IsString via Identifier
   deriving Identifiable via Identifier
   deriving Semigroup via Identifier
   deriving Monoid via Identifier
   deriving Pretty via Identifier
 
--- | ToIdentifier for struct field names, e.g., @len@
-newtype FieldId = FieldId Identifier
+-- | de Bruijn index for type variables, e.g., @'0@
+newtype TVarIdx = TVarIdx Int
   deriving Eq
   deriving Ord
-  deriving Show via Identifier
-  deriving ToIdent via Identifier
-  deriving IsString via Identifier
-  deriving Identifiable via Identifier
-  deriving Semigroup via Identifier
-  deriving Monoid via Identifier
-  deriving Pretty via Identifier
+  deriving Typeable
+  deriving Data
 
--- | A name to be bound; 'Nothing' represents a wildcard, e.g., @let _ = ...@
+instance Show TVarIdx where
+  show (TVarIdx i) = "'" ++ show i
```

Comment:
Could we change this string representation to something else? `'0` is pretty nice, but then `brittany` will throw a ParseError when I use it to pretty print the AST, since it cannot handle the single `'` before the index and will consider it to be an unbalanced quotation case. (That's why I changed it to `show (TVarIdx i) = "t" ++ show i` in my type inference PR).

Reply:
You raise a good point, but I would actually encourage you to just remove these altogether in your branch, since there's really no need to represent type variables as integers (just makes output harder to read). I kept this in here in case it was still being used elsewhere)